### ISSUE TITLE
dark mode

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#10189]
Dark mode should be implemented for:
- when someone is using Cockpit at night or in a dark room
- accessibility reasons (low to medium migraine, various vision issues)
- (as a preference for) people who like dark mode

---

As Cockpit has a lot of moving parts, it's important that we align everything toward one release so that Cockpit can be completely dark without flashes of light or pages that are bright

## Proof of concept

- PF4 dark mode for a subset of pages, to see what happens
  - are we able to add it to `page.scss`?

## Minimum implementation for a Cockpit release

Dark mode works:
- based on system theme
- across all of base Cockpit
- in all of our own plugins
- at the login screen
- in Cockpit Client
  - respects system dark mode with the UI (titlebar)
  - passes system dark mode to Cockpit (unless WebKit does this automatically, which it may)
- in graphs
- in the metrics page 
- in terminal, as there are themes, so dark doesn't get inverted to a light theme
- in non-PF4 pages (filter hack)
- in remote older versions of Cockpit?

It doesn't make sense to have a half-baked dark mode (unless we have some opt-in beta hack), so these should really land to be available to users in the same release.

## Meanwhile

Once we have the start of an implementation that is working, we should reach out to other teams who work in the Cockpit ecosystem, such as 45Drives (who work on several 3rd party Cockpit plugins), SUSE, and the installer teams (at Red Hat and SUSE).

## Possible follow-ups

- Setting
- Light mode (where the navigation and top bar are also light)

**Comment by garrett** [comment]
I can't tease a dark mode without screenshots. So here are some screenshots. Remember, this is the _only 1 line of CSS_ proof-of-concept.

## System overview
![screenshot_2018-10-01 system - sunny](https://user-images.githubusercontent.com/10246/46278581-d83ccf00-c566-11e8-83c9-ed11c565ee4c.png)

## Storage page
![screenshot_2018-10-01 storage - sunny](https://user-images.githubusercontent.com/10246/46278585-dbd05600-c566-11e8-9628-bd751a1a718e.png)

## Dialog on the storage page
(Here's where the fix for item 2 would be needed.)
![screenshot_2018-10-01 storage - sunny 1](https://user-images.githubusercontent.com/10246/46278592-df63dd00-c566-11e8-86b4-71eeaa064fde.png)

**Comment by garrett** [comment]
## Mobile

(I scaled this down so it wouldn't be super-huge as a preview here.)

![screen shot 2018-10-01 at 10 46 25](https://user-images.githubusercontent.com/10246/46278770-7466d600-c567-11e8-84a6-c95650e337bf.png)

**Comment by aaronstuder** [comment]
Very Nice!

**Comment by rhwood** [comment]
See also https://stackoverflow.com/a/51799496/176160 It is likely there will soon be some support for determining if the browser is running in dark mode.

**Comment by garrett** [comment]
Firefox 67 landed this week and [now includes dark mode support](https://hacks.mozilla.org/2019/05/firefox-67-dark-mode-css-webrender/).

- On OSes that have a toggle, it supports the toggle.
- On Linux, it checks the GTK foreground and background colors and infers a theme
- On all OSes, there are overrides

OS-based dark mode in browsers ([prefers-color-scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)) is only supported in Firefox and Safari. Chrome support is forthcoming (and should land some time this year).

**Comment by garrett** [comment]
I'm currently working on overhauling all our colors in issue https://github.com/cockpit-project/cockpit/issues/11913.

Once done, this should allow us to support dark mode in a cleaner and more performant way, by redefining all the colors. (That said, I didn't notice performance issues. But it probably would have at least a little impact to flip the colors around.)

The method used above could be useful for parts of Cockpit that might not be easy to re-theme (such as the graphs), other top-level add-ons (such as Composer) or remote machines running old versions of Cockpit. The latter two exapmples would probably need additional work — perhaps by adding a special class in an iframe to enable color-variable-based dark mode and then defaulting to the above inversion method otherwise (if the class does not exist). We'd want to make sure there isn't a jarring white background in dark mode, in other words, and would need to mix-and-match techniques to get it done.

**Comment by garrett** [comment]
The color reworking was absorbed in to an overall restyling of Cockpit in https://github.com/cockpit-project/cockpit/pull/11987.

Not all the colors have been ported over to variables, but most have.

Adding a proper dark mode would help spot other colors that have not been adjusted.

Meanwhile, here's a great article on the whens, whys, & hows of dark mode:
https://web.dev/prefers-color-scheme/

**Comment by Dorson** [comment]
Coded a flexible and working solution for the Dark Mode.

Demo in here:
https://dorson.github.io/CSS-Dark-Mode-and-color-switch/

Works with basic CSS variables + short JavaScript.

Can be used to define multiple color themes in one CSS file or other dynamic style setting options like text size, etc...

Have fun using my code !

**Comment by garrett** [comment]
More dark mode info, BTW:
https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/

**Comment by garrett** [comment]
@Dorson: Thanks for the link. However, Cockpit is pretty complex, being a mix of PF3, PF4, and a lot of custom code. We'd have to have our own implementation of dark mode. PF4 already has some preliminary dark widget support, but that would only get us so far. Whatever is made for Cockpit would have to handle everything. We'd probably have to use more than one technique. (Like an optimized path when possible and possibly even a fallback one that is slightly less performant for legacy code and third-party extensions that don't directly support a dark mode.)

Plus, there are various design considerations to take in mind when developing a dark mode (more than inverting + shifting hue), as mentioned in the link I just shared above.

Glad you know that you're interested in dark mode though! The higher the want for it, the more likely we can get to it sooner than later (if at all).

**Comment by garrett** [comment]
Copy/pasted from https://github.com/martinpitt/performance-graphs/pull/64#issuecomment-702187607:

---

Ideally, PatternFly would have a class for dark mode. IIRC, they would enable it at the top of the page _or_ at a specific widget level.

So we could do something like this in SCSS:

```scss
.foo {
   color: var(--some-color);
  .dark-mode & {
     --some-color: #fff;
     background: red;
  }
}
```

And SASS would unwrap that as:

```scss
.foo { ... }
.dark-mode .foo { ... }
```

And the specificity of the dark mode overrides would win out. It would also be afterward too. :wink:

FWIW: PF would probably go with `.pf-m-darkmode` or something such.

---

Also: If dark mode is applied to the same widget, then we'd do this instead:

```scss
.foo {
   &.dark-mode {}
}
```

But, in most cases, we'd want dark mode at the top of the page and have everything inherit.

**Comment by garrett** [comment]
Ideally, we'd try to use PF4's dark mode (with it lands) and have a fallback for non-PF with the example I outlined above (in which we'd probably also drop or adjust the dropshadows so they're not "dropglows").

And we'd have a little bit of CSS like the above comment for one-off overrides where necessary.

**Comment by martinpitt** [comment]
Nobody else asked for this as far as I can remember, but this issue has a PoC and is clearly stated, so if anyone wants to have a go at this, it's still relevant.

**Comment by garrett** [comment]
Removing and updating the colors we use (#15183) was a walk in the direction of this issue. And PatternFly is working on support. Once PF lands dark mode, we can try to see what happens when enabling it.

I'm still certain we'll need a PF4 native solution _and_ a compatibility one (based on this proof of concept) as well, as mentioned above.

**Comment by garrett** [comment]
> Nobody else asked for this as far as I can remember

It's an accessibility feature and a user preference. Over the past few years, it's been showing up almost everywhere (with good reasons).

[elementary OS](https://elementary.io/) (opinionated Linux-based OS that has an app store) founders have a good writeup about implementing a dark style preference... it's a nice read and makes a lot of great points (especially around accessibility) and is mostly not elementary-related:
https://blog.elementary.io/the-need-for-a-freedesktop-dark-style-preference/

Meanwhile, coincidentally, I have my blinds mostly closed, sitting in a dark room, actually using dark "mode"\* today as I have a bit of a headache and I want to keep working and try to not go into massive migraine mode. (It's important to me and _very likely_ a lot of our users.)

_\*  Technically, it's the `adwaita-dark` theme in gnome-tweaks. Firefox recognizes the light text and dark background, skins itself into a dark mode and then sets its own internal settings to to prefers-dark. This then causes GitHub and several other websites to flip to dark mode too._

**Comment by jchristi** [comment]
@garrett Have you tried the [Dark Reader](https://darkreader.org/) browser extension? I can't live without it!

**Comment by garrett** [comment]
@jchristi: Yeah, Dark Reader is great! I use it on my phone especially (via Firefox on Android). 

It doesn't really work with Cockpit, however, as we have a bunch of CSP (Content Security Policy) protections on which (rightfully) get in the way. We'll need to have our own solution for dark mode.

**Comment by garrett** [comment]
Related, upstream PF dark theme proof of concept: https://github.com/patternfly/patternfly/pull/4203

**Comment by septatrix** [comment]
Would definitely like the modern look of Patternfly 4

**Comment by garrett** [comment]
@septatrix: Cockpit has had the modern look of PatternFly 4 for a few years already... If your version of Cockpit is still using PatternFly 3, then you might want to consider updating. 

If you're on an older version that ships in Debian Stable or Ubuntu LTS, we have backports instructions @ https://cockpit-project.org/running.html#debian (RHEL ships newer Cockpits and we have [a COPR](https://copr.fedorainfracloud.org/coprs/g/cockpit/cockpit-preview/) available for those who want the absolute latest in the form of a preview.)

Cockpit is currently at version 263 as of right now.

As for the next iteration of PatternFly 4, PatternFly itself hasn't finalized that yet. But it's probably when upstream dark support will land, and then we could use it in Cockpit (although with a bit of effort on our end too, for sure).

**Comment by septatrix** [comment]
Oh yes please ignore me. I only check the dashboard sparingly and confused it with FreeIPA and the Ansible Galaxy homepage :D

**Comment by Archef2000** [comment]
I would like to know how to implement it on Cockpit @garrett from your first comment
```
.area-ct-content {
   filter: hue-rotate(180deg) invert(150%) brightness(1.5);
}
```
if not directly maybe trough Nginx

**Comment by garrett** [comment]
@Archef2000: That won't work, as Cockpit has changed a ton since this issue was first opened. We're using PatternFly 4 now, instead of PatternFly 3. And we're using many more PatternFly widgets instead of a lot of custom ones too.

An updated version of the proof of concept hack would be to add something like this, probably in `page.scss`:

```scss
.pf-c-page {
  filter: hue-rotate(180deg) invert(150%) brightness(1.5);
}
```

However, this is not the right way to do things:
- It affects performance (in _most_ cases: very slightly)
- It causes icons to be too light
- It doesn't work with the native PatternFly dark mode (which is upcoming)
- Shadows become odd glows
- Backdrops behind modals are lighter instead of darker, causing a blast of lightness in dark environments
- It doesn't respect the user's `prefers-color-scheme` settings (which has landed in all browsers since I first opened this issue)

Most of these can be worked around with additional CSS. And we'd need to do something like this anyway without PatternFly 4's native dark mode setting, for pages that don't support that.

But it's finally probably around the time where we can start looking into adding some kind of dark mode.

We'd want to have fallback dark mode first, then patch in PF4 dark mode with a way to override the fallback (most likely using `:not()`), as having partial dark mode support does not make sense. (It's a bad experience to get some pages and some widgets in dark mode with others not.

**Comment by Archef2000** [comment]
Where can i find the page.css ?

**Comment by garrett** [comment]
Note: a _better_ fallback might be to redefine the PF color variables. (And I think this is what PF does for their implementation?) But we'd probably still need a hacky fallback too, as not every page in Cockpit would support it. In those cases, we might even need to attach the CSS with a more generalized selector, like `body`; perhaps `body:not(.pf-c-page)` (which would say pages that do NOT use PatternFly 4). (We'd need example pages.)

**Comment by garrett** [comment]
> Where can i find the page.css ?

It's `page.scss` in the Cockpit source. You'd need to build Cockpit for that. But, again, it's a complete hack and would need more work done.

Otherwise, injecting the CSS in an already-compiled Cockpit somewhere would probably work... but I think you'd need to do that on every page's compiled CSS file?

There are also some browser extensions that provide a means of adding CSS to pages, such as Stylus, and others that provide a hacky dark mode, like Dark Reader... but these kinds of extensions may have issues with Cockpit's CSP (content security policy) settings.

**Comment by Archef2000** [comment]
And in Nginx

**Comment by garrett** [comment]
> And in Nginx

I don't know about Nginx. That's completely out of scope of implementing a dark mode in Cockpit.

When I do use a proxy in the front, I use Caddy... Sorry; I can't help further with Nginx. A quick search on DuckDuckGo showed this result that might help you: https://forum.level1techs.com/t/infrastructure-series-use-nginx-to-inject-css-themes/174165

I can _eventually_ implement dark mode directly in Cockpit. I know @jelly has also expressed interest in this. We want to try to tackle it "soon", but we're both juggling many other tasks in Cockpit (and Cockpit-Podman, Cockpit-Machines, etc.), so I can't give you an estimate aside from our wish to try to get it done sooner than later.

**Comment by garrett** [comment]
Meanwhile, PatternFly has an issue open about guidance for dark theme implementations for software that uses PatternFly 4, such as Cockpit (but this is a TODO item):
- https://github.com/patternfly/patternfly-design/issues/1152

And they have a meta-issue ("epic") about their dark theme implementation @:
- https://github.com/patternfly/patternfly-design/issues/754

Where they say:

> dark theme is not yet available for general use. We currently have an alpha version that we are testing with selected projects and don't have any firm date on when this might be fully supported

---

Additional notes: As the PF dark theme can be implemented at a page or a widget level, for a fallback theme to work properly, we'd want to use a way that cascades down as to not invert an already-dark mode. That is, we would want to probably use something like the CSS custom property trick to have a cascading boolean in CSS which would let us either apply or skip application of any dark mode settings, so we don't accidentally apply more than 1 dark mode change at the same time.

https://css-tricks.com/the-css-custom-property-toggle-trick/

So, in order of ideal implementation to less ideal, it would be something like:

1. PatternFly 4's dark theme setting
2. CSS variable redefining (where PF4's dark theme is not supported, which in some case, might even be some PF4 widgets that might not have their own dark mode setting yet... and in other cases, Cockpit's custom widgets)
3. CSS filter to invert (especially on pages where PF4 does simply not exist... like in some 3rd party apps / add-ons for Cockpit)

If 1 is active, we don't want 2 or 3. If 2 is active, we don't want 3 (and 1 wouldn't be possible if 2 is active). If 3 is active, neither 1 nor 2 would be possible.

So, in other words: Fallbacks on fallbacks.

Also, this wouldn't be able to apply to remote servers, unless the remotes would also be dark mode capable (that is: on whatever version Cockpit would have a dark mode).

Cockpit is much more complicated than any other app using PatternFly 4, as we have iframes with embedded pages that might or might not be developed by us — and that might or might not even be made using PatternFly. And, we can also have remote servers with different versions of Cockpit too.

**Comment by garrett** [comment]
A for a CSS variable redefinition method, a good first pass to figure out which colors would need to be reassigned (when dark mode is on) would be to view the PF color page (https://www.patternfly.org/v4/guidelines/colors/) and use a simple invert on the `html` or `body` tag, such as: 'filter: invert() hue-rotate(180deg);`

It would then look like this preview:

![image](https://user-images.githubusercontent.com/10246/165295330-968f0c17-b7d0-4345-9405-fcc446a96dc2.png)

And then someone (probably me) would have to redefine all the PF color variables to the color picker version of the colors shown there. Some would have to be tweaked with something like the above-mentioned filter (`filter: hue-rotate(180deg) invert(150%) brightness(1.5);`) — this would most notably be the the grey colors, to provide a bit more contrast.

This would provide a way to have a dark mode based on color redefinition, like I mentioned as the first fallback (# 2 in the above list), and we already have a hacky approach for a # 3 (the CSS filter with some additional CSS to correct some of the things it gets wrong).

Then, the largest task would be integrating all of this and making sure the most preferable approach is used, with fallbacks where necessary. (We'd also want to adjust the CSS variable redefinition to whatever PF winds up using in their implementation too. But the approach I mention in this comment is a good first pass for a starter implementation.)

**Comment by garrett** [comment]
I've talked with some of the PF devs about enabling dark mode in PatternFly and got a response. Here's an excerpt:

> The TLDR is that you import `@patternfly/patternfly/patternfly-theme-dark.css` directly after the base PF styles, so it can correctly override those as needed but not conflict with any additional overrides you may load after, and then apply the .pf-theme-dark class to the `<html>` tag to toggle the dark theme styles on.  

There's also an internal WIP document with more details, including gotchas. I'm happy to share it with anyone on the team. (It's not public, so only RHers can access it for now, but some form of it will be published publicly once PF makes dark mode official for everyone. They want to iron out bugs first before officially turning it on for everyone, hence the semi-closed beta for now.)

For everyone following along: This, of course, would have to be done in the source files before compilation. There isn't a way to enable it in Cockpit builds, and it will require some work on Cockpit to properly enable it.

**Comment by garrett** [comment]
I took a screenshot of Cockpit Client with the login page

![Screenshot from 2022-05-23 16-10-36](https://user-images.githubusercontent.com/10246/169843604-793b9c0d-9eb2-49a9-822c-28f11e9f3f06.png)

And made a dark mode mockup using the colors from the current PatternFly dark mode. It's not meant to be pixel perfect; it's just approximate. Colors and the general feel are the important part of this mockup.

![login-dark](https://user-images.githubusercontent.com/10246/169843801-3408b88a-5877-4f7d-87ec-c7fb8f496034.png)

I've added a black screen at 75% behind the login, but above the logo, to make the background wallpaper and the OS logo darker. This also causes the login entry to still "pop" visually with a but more contrast between the login and the wallpaper. It also makes the screen overall darker, which is the point of dark mode, right? :wink:

**Comment by garrett** [comment]
@Archef2000: No files. This is for reference. 

I was going to add dark mode to the login page, but @allisonkarlitskaya wants to pick it up. She'll implement it and I'll look it over. We intend on doing this before or by Wednesday. (No promises, of course!)

Meanwhile, I looked into implementing dark mode for Cockpit itself, first as a hack (mainly for debugging purposes) and still hit some walls. It'll take some more effort for Cockpit itself.

But @allisonkarlitskaya [added dark mode for Cockpit Client last release](https://cockpit-project.org/blog/cockpit-269.html) and we'll likely add the login page dark mode for the upcoming release. It'll probably take longer for dark mode for Cockpit (and the various pages) itself, as that's much more complex.

**Comment by Archef2000** [comment]
> @Archef2000: No files. This is for reference.
> 
> I was going to add dark mode to the login page, but @allisonkarlitskaya wants to pick it up. She'll implement it and I'll look it over. We intend on doing this before or by Wednesday. (No promises, of course!)
> 
> Meanwhile, I looked into implementing dark mode for Cockpit itself, first as a hack (mainly for debugging purposes) and still hit some walls. It'll take some more effort for Cockpit itself.
> 
> @allisonkarlitskaya [added dark mode for Cockpit Client last release](https://cockpit-project.org/blog/cockpit-269.html) and we'll likely add the login page dark mode for the upcoming release. It'll probably take longer for dark mode for Cockpit (and the various pages) itself, as that's much more complex.

But how did you get the login page etc. dark?

**Comment by jelly** [comment]
I noticed running the following JS in your developer console on https://www.patternfly.org/v4/components/alert changes it to the dark theme:

```
document.documentElement.classList.toggle('pf-theme-dark');
```

**Comment by garrett** [comment]
I rebased this branch earlier today. There were some conflicts, but I think I solved them all.

After that, I've been hunting down the way dark mode should look. The best overview of all the components seems to be at https://patternfly-react-pr-7644.surge.sh/components/form — be sure to _manually_ enable the toggle in the top-right.

I'm not sure why this isn't on PatternFly.org by default.

Anyway, this is what I'm using for reference now, to figure out which parts of Cockpit are broken in dark mode and what they should look like with a fix.

**Comment by septatrix** [comment]
> Anyway, this is what I'm using for reference now, to figure out which parts of Cockpit are broken in dark mode and what they should look like with a fix.

First and foremost the navbar in PF4 seems to be broken in dark mode...

![image](https://user-images.githubusercontent.com/24257556/193871947-f699110b-5ba1-4c7f-b6a9-78674e6e2d02.png)

I am not sure if this might be fixed in newer versions because the main docs do not seem to have the dark mode toggle for now so I did not want to file an issue yet

**Comment by garrett** [comment]
latest PR branch from PF has a bunch of things fixed in dark mode

Warning buttons are legible with correct contrast again
https://patternfly-pr-5137.surge.sh/components/button

Search inputs are finally styled correctly
https://patternfly-pr-5137.surge.sh/components/search-input

And a bunch of other subtle things.

We need to upgrade to the latest PF to properly continue work.

**Comment by jelly** [comment]
https://github.com/cockpit-project/cockpit/pull/17321 now merged with dark mode.

**Comment by septatrix** [comment]
> #17321 now merged with dark mode.

That PR only implement a dark theming for the GTK wrapper but not the web interface itself. This is only a minute fraction of what this issue is encompassing and should therefore very much remain open.

**Comment by jelly** [comment]
See the latest release, dark mode was fully implemented https://cockpit-project.org/blog/cockpit-279.html

**Comment by jelly** [comment]
@Archef2000 do you have a dark theme enabled in your environment? What browser are you using?

**Comment by Archef2000** [comment]
I tested Chome desktop and Firefox and https://cockpit-project.org/running.html says that my browser is compatible.

**Comment by Archef2000** [comment]
![Screenshot 2022-11-09 145611](https://user-images.githubusercontent.com/69057532/200852339-3f526b27-0eb6-4397-b84b-6de3ff89bab4.png)

**Comment by Archef2000** [comment]
No it has not. The top where Assministrative access, Help and Session are localted was white. 
And this is the new login screen.
![Screenshot 2022-11-09 162109](https://user-images.githubusercontent.com/69057532/200870190-4d1f77eb-6ff0-4465-a62e-014b150ca787.png)

**Comment by jelly** [comment]
@Archef2000 which Cockpit version is this?

**Comment by Archef2000** [comment]
Latest backport like the install docs say.

**Comment by Archef2000** [comment]
Oh my bad i did not check what i installed how could i install v 279

**Comment by alanmilinovic** [comment]
It is not yet there, currently in unstable repo.

Wait for it.
https://tracker.debian.org/pkg/cockpit

**Comment by alanmilinovic** [comment]
I guess in couple of days.

**Comment by garrett** [comment]
Cockpit has had dark mode introduced in parts. First was [Cockpit Client gaining dark mode support in version 269](https://cockpit-project.org/blog/cockpit-269.html) and [the login screen in version 272](https://cockpit-project.org/blog/cockpit-272.html#login-dark-mode), as both are fairly standalone compared to the rest of Cockpit.

What you're seeing in a slightly older version of Cockpit (272 - 278) is the login screen's dark mode support.

Finishing up dark mode support for Cockpit when you're logged in required a bunch of extreme debugging of CSS. We're also one of the first projects using PatternFly to use dark mode support, which is still a little rough around the edges (so we had to debug and workaround a few small issues with it too).

We are also strongly considering having a way in Cockpit's UI to switch between auto (using system's setting, which is what we're doing right now as of 279), light, and dark. I have something in mind and am hoping to make a design for this today. Hopefully we can get it implemented soon.

Basically having some kind of switch would allow you to override whatever light/dark mode setting your system sends to the browser which sends to Cockpit. This would be especially useful for buggy browsers* that sometimes have problems detecting dark mode.

\* I'm specifically referring to Chrome, Chromium, Edge, and other Chromium-based browsers on Linux. They all have issues, especially on some desktops versus others. Firefox and WebKit browsers (GNOME Web / Epiphany, Cockpit Client, Safari) work well everywhere.

**Comment by alanmilinovic** [comment]
How to have dark mode applied to plugins? I tested "Pinger" example as well and it is with white background.

**Comment by arch-user-france1** [comment]
> * I'm specifically referring to Chrome, Chromium, Edge, and other Chromium-based browsers on Linux. They all have issues, especially on some desktops versus others. Firefox and WebKit browsers (GNOME Web / Epiphany, Cockpit Client, Safari) work well everywhere.

**clarify some things**
1. Don't call those browsers buggy, it's mostly snap failing or similar software
2. In Firefox, white theme is forced when "privacy.resistFingerprinting" in about:config is set to true
3. In Brave, white theme is forced when extreme fingerprint blocking is enabled, but that can be disabled in chrome:config for every site

**Comment by Archef2000** [comment]
For anybody following this issue cockpit 279 is now in debian backports.

**Comment by alanmilinovic** [comment]
Since yesterday. Can anyone explain how to have dark mode for plugins? Pinger example is also white, everything else is dark.

**Comment by Archef2000** [comment]
The plugins will need to be updated and add support for darkmode.

**Comment by alanmilinovic** [comment]
The question is how to update plugin to support darkmode? 😀

**Comment by Archef2000** [comment]
Then im sorry that i did not understand your question .

**Comment by garrett** [comment]
> Don't call those browsers buggy, it's mostly snap failing or similar software

I meant the experience of using those browsers on Linux specifically is buggy. There's literally a Chromium bug filed in their issue tracker that's been open for a few years that's unfixed, so it really does look like a bug in Chromium. And it works on some desktops, but not others (such as KDE and GNOME, the two major desktops). I'm pretty sure this means it qualifies as a bug.

FWIW: It's a problem with the Flatpak version of Chrome (and other Chromium browsers) too, not just the snap. Meanwhile, Firefox and Epiphany handle it seamlessly, wherever, including system-installed _and_ Flatpak. :shrug: 

> In Firefox, white theme is forced when "privacy.resistFingerprinting" in about:config is set to true
> In Brave, white theme is forced when extreme fingerprint blocking is enabled, but that can be disabled in chrome://config for every site

All the more reason to have a manual switch, which I did the designs for this past week. Hopefully we'll get it in the next version of Cockpit (no promises). I _think_ it's pretty simple to implement, so it's likely.

> The question is how to update plugin to support darkmode? :grinning: 

Yes, good point. We need to update that as well.

And third party pages too, most notably @45Drives. And also the new Fedora/CentOS/RHEL Anaconda and SUSE installers, both of which are based on Cockpit.

So while dark mode is "done" in Cockpit itself, there's still more work to be done.

**Comment by jelly** [comment]
> The question is how to update plugin to support darkmode? grinning

The plugin developer needs to update to the latest Patternfly as every plugin ships it's own CSS/HTML.

**Comment by garrett** [comment]
> What is the pinger plugin?

@jelly: It's our example plugin.

- https://cockpit-project.org/blog/cockpit-starter-kit.html
- https://cockpit-project.org/blog/creating-plugins-for-the-cockpit-user-interface.html

In other words, https://github.com/cockpit-project/starter-kit needs an update too.

**Comment by alanmilinovic** [comment]
> > The question is how to update plugin to support darkmode? grinning
> 
> The plugin developer needs to update to the latest Patternfly as every plugin ships it's own CSS/HTML.
> 
> What is the pinger plugin? I can't find it under https://cockpit-project.org/applications.html

It is just an example within project.
https://github.com/cockpit-project/cockpit/tree/main/examples/pinger
